### DISCUSSION TITLE
[不算Bug的问题] missing or invalid media type header 报错

### DISCUSSION THREAD

**Discussion by pmkol** [IrineSistiana/mosdns#522] [Q&A]
### 在提交之前，请确认

- [X] 我已经尝试搜索过 Issue ，但没有找到相关问题。
- [X] 我正在使用最新的 mosdns 版本(或者最新的 commit)，问题依旧存在。
- [X] 我仔细看过 wiki 后仍然无法自行解决该问题。
- [X] 我非常确定这是 mosdns 核心的问题。(如果是通过第三方衍生软件使用 mosdns 核心，不确定问题源头时，请先向衍生软件开发者提交问题。)

### mosdns 版本

4.x

### 操作系统

centos

### Bug 描述和复现步骤

貌似用浏览器请求一些DOH链接时，就会在日志里出现这些报错，我记得是mosdns做了什么判断，当缺少什么条件时，就会报错missing or invalid media type header

当一些人尝试用浏览器返回DOH结果测试时，就会导致这种报错产生，有时候量会很大，影响看日志数据。

想了解这个报错是不是能加个参数控制一下，比如关闭时，用浏览器也能访问到DOH。

又或者这个参数是通过什么判断的，是否可以通过NGINX做前端通过UA或者其他什么条件限制掉。

### 使用的配置文件

```yaml
-
```

**Comment by urlesistiana** [comment]
DoH 要求请求必须有个 `Accept` 头，值为 `application/dns-message`

**Comment by pmkol** [comment]
nginx 有什么办法能限制么？
或者是否可以在mosdns配置里加个参数，关闭这个限制

**Comment by emptyteeth** [comment]
可以试试让nginx加上这个头
proxy_set_header Accept application/dns-message;

**Comment by pmkol** [comment]
> 可以试试让nginx加上这个头 proxy_set_header Accept application/dns-message;

尝试了这个方法，但是会出现另外一种错误
2022-10-06T04:41:24.840+0800	warn	http_handler/handler.go:81	invalid request	{"from": "127.0.0.1:38042", "method": "GET", "url": "/dns-query?123", "error": "no dns parameter"}

@urlesistiana 是否可以加一个参数，限制这个报错，因为对方可以利用这个漏洞就行攻击，刷爆mosdns的日志

**Comment by urlesistiana** [comment]
> 是否可以在mosdns配置里加个参数，关闭这个限制

这个头是 DoH 规范强制要求的。

> 可以试试让nginx加上这个头
proxy_set_header Accept application/dns-message;

不推荐。

你可以让 nginx 负责检查这个头。没有这个头或者有问题的直接转 403。

**Comment by pmkol** [comment]
> > 是否可以在mosdns配置里加个参数，关闭这个限制
> 
> 这个头是 DoH 规范强制要求的。
> 
> > 可以试试让nginx加上这个头
> > proxy_set_header Accept application/dns-message;
> 
> 不推荐。
> 
> 你可以让 nginx 负责检查这个头。没有这个头或者有问题的直接转 403。
```
if ($http_accept != "application/dns-message") {
  return 412;
}
```
有通过这个方法去限制，但从CDN的报表观察，有大概10%的请求，都会报错，而这些请求在mosdns的wran日志中没有报错提示，日志那里是不是可以加个参数，比如用户可以指定屏蔽某种类型的报错，我的DOH请求量较大，一天约1500万次了，日志因为DOH这个头报错很困扰

**Comment by urlesistiana** [comment]
> 比如用户可以指定屏蔽某种类型的报错，

不行，太复杂。mosdns 的日志只负责报错。过滤的任务最好交给其他 log 管理平台。

**Comment by pmkol** [comment]
> > 比如用户可以指定屏蔽某种类型的报错，
> 
> 不行，太复杂。mosdns 的日志只负责报错。过滤的任务最好交给其他 log 管理平台。

我上面ngninx的写法是有什么地方不对么？为什么会拦截掉很多正常请求呢？
